capsules/src/nrf51822_serialization.rs

```diff
@@ -244,24 +281,26 @@ impl uart::ReceiveClient for Nrf51822Serialization<'_> {
 
         self.active_app.map(|appid| {
             let _ = self.apps.enter(*appid, |app, _| {
-                app.rx_buffer = app.rx_buffer.take().map(|mut rb| {
+                let len = app.rx_buffer.mut_map_or(0, |rb| {
```

Comment:
I believe the `len` variable and the `mut_map_or` and later `map_or` are wrong here.

The comment below says `Note: This indicates how many bytes were received by hardware, regardless of how much space (if any) was available in the buffer provided by the app.`

But with the current code, if the `rx_buffer` is missing it will return `0` instead, and if it's too short, the later `max_len` will return the app's `rx_buffer` length, not the returned hardware length.

I think the fix is simply to remove the `len` variable, remove all of the `_or`'s here, and pass `rx_len` in the callback below. I'm not going to push that fix without consensus though, as I think we'd be going back and forth if I did that..

Reply:
Good point -- then we should pass both. I assume these semantics are there for the peculiarities of the nRF serialization library? I'd therefore recommend this become

```rust
app.callback.schedule(4, rx_len, len);
```

Having implemented a bunch of these capsules, I've come to think that a simple `map` is dangerous; the situations in which you want to plug away when there's nothing there is pretty narrow. The API you describe, in which the byte count is the number received by the hardware, rather than what's in the buffer, is fine as long as there is always a buffer (of known length) there. But this `map` indicates that there might not be; calling `map` prevents you from being able to detect an error case. It might be that we do not want to dig into the nRF code to handle this case right now, but I think it doesn't speak well for the kernel if we make it impossible to handle. `map_or` has the advantage that it forces you to think about this -- what if the buffer isn't there?